File: MATLAB/astar.m
function path = astar(grid, start, goal)
    % A* Pathfinding Algorithm in MATLAB
    % Inputs:
    %   grid  - Matrix representing the map (0: path, 1: obstacle)
    %   start - Starting point in format [row, col]
    %   goal  - Goal point in format [row, col]
    % Output:
    %   path  - List of points representing the shortest path from start to goal
    
    [rows, cols] = size(grid);
    openSet = [start, 0, heuristic(start, goal), heuristic(start, goal)];
    closedSet = zeros(size(grid));
    cameFrom = cell(size(grid));
    
    % A* Loop
    while ~isempty(openSet)
        [~, idx] = min(openSet(:, 4)); % Find the node with lowest f cost
        current = openSet(idx, 1:2);
        
        if isequal(current, goal)
            path = reconstruct_path(cameFrom, current, start);
            return;
        end
        
        openSet(idx, :) = []; % Remove current from openSet
        closedSet(current(1), current(2)) = 1;
        
        % Check all neighbors (4-directional movement)
        for d = [-1 0; 1 0; 0 -1; 0 1]'
            neighbor = current + d';
            if neighbor(1) < 1 || neighbor(1) > rows || neighbor(2) < 1 || neighbor(2) > cols
                continue;
            end
            if grid(neighbor(1), neighbor(2)) == 1 || closedSet(neighbor(1), neighbor(2)) == 1
                continue;
            end
            
            tentative_gScore = heuristic(current, start) + 1;
            if ~ismember(neighbor, openSet(:, 1:2), 'rows')
                cameFrom{neighbor(1), neighbor(2)} = current;
                fScore = tentative_gScore + heuristic(neighbor, goal);
                openSet = [openSet; neighbor, tentative_gScore, heuristic(neighbor, goal), fScore];
            end
        end
    end
    error('No path found');
end

function h = heuristic(point, goal)
    % Heuristic function using Manhattan distance
    h = abs(point(1) - goal(1)) + abs(point(2) - goal(2));
end

function path = reconstruct_path(cameFrom, current, start)
    % Reconstruct path from goal to start
    path = current;
    while ~isequal(current, start)
        current = cameFrom{current(1), current(2)};
        path = [current; path];
    end
end

File: Python/astar.py
import heapq

class Node:
    def __init__(self, x, y, g=0, h=0):
        self.x = x
        self.y = y
        self.g = g  # Movement cost from start to current node
        self.h = h  # Heuristic cost to reach goal
        self.f = g + h  # Total cost
        self.parent = None  # Tracks the path

    def __lt__(self, other):
        return self.f < other.f

def heuristic(node, goal):
    """Calculate the heuristic based on Manhattan distance."""
    return abs(node.x - goal.x) + abs(node.y - goal.y)

def astar(start, goal, grid):
    open_set = []
    closed_set = set()
    heapq.heappush(open_set, (0, start))
    
    while open_set:
        _, current = heapq.heappop(open_set)
        
        if (current.x, current.y) == (goal.x, goal.y):
            path = []
            while current:
                path.append((current.x, current.y))
                current = current.parent
            return path[::-1]  # Return reversed path
        
        closed_set.add((current.x, current.y))
        
        # Explore neighbors
        for dx, dy in [(-1, 0), (1, 0), (0, -1), (0, 1)]:  # Up, Down, Left, Right
            nx, ny = current.x + dx, current.y + dy
            if 0 <= nx < len(grid) and 0 <= ny < len(grid[0]) and grid[nx][ny] == 0:
                neighbor = Node(nx, ny, current.g + 1)
                neighbor.h = heuristic(neighbor, goal)
                neighbor.f = neighbor.g + neighbor.h
                neighbor.parent = current
                
                if (neighbor.x, neighbor.y) in closed_set:
                    continue
                
                heapq.heappush(open_set, (neighbor.f, neighbor))
    
    return None  # No path found
